sklearn/preprocessing/data.py

```diff
@@ -1901,3 +1902,200 @@ def transform(self, X):
         """
         return _transform_selected(X, self._transform,
                                    self.categorical_features, copy=True)
+
+
+class CountFeaturizer(BaseEstimator, TransformerMixin):
+    """Adds a feature representing each feature value's count in training
+
+    Specifically, for each data point X_i in the dataset X, it will add in
+    a new feature vector count_X_i to the end of X_i where count_X_i is the
+    number of occurences of X_i in the dataset X given the equality indicator
+    'inclusion'.
+
+    If a y argument is given during the fit step, then the
+    count in the transform step will be conditional on the y.
+
+    This preprocessing step is useful when the number of occurences
+    of a particular piece of data is helpful in computing the prediction.
+
+    Parameters
+    ----------
+    inclusion : 'all', set, list, or numpy.ndarray
+        The inclusion criteria for counting
+
+        - 'all' (default) : Every feature given is counted
+        - collection of indices : Only the given collection of features is
+        counted
+
+    Attributes
+    ----------
+    count_cache_ : defaultdict(int)
+        The counts of each example learned during 'fit'
+        For example (X_i, y_i), the count can be accessed with 
+        count_cache_[X_i][y_i]
+
+    y_set_ : list of (index, y) tuples
+        An enumerated set of all unique values y can have
+
+    col_num_X_ : int
+        The number of columns of 'X' learned during 'fit'
+
+    col_num_Y_ : int
+        The number of clumns of 'y' learned during 'fit'
+        If 0, then the fit is not conditional on the y given
+
+    Examples
+    --------
+    Given a dataset with two features and four samples, we let the transformer
+    find the number of occurences of each data point in the dataset
+
+    >>> from sklearn.preprocessing.data import CountFeaturizer
+    >>> X = [[0], [0], [0], [0], [1], [1], [1], [1]]
+    >>> y = [0, 1, 1, 1, 0, 0, 0, 0]
+    >>> cf = CountFeaturizer()
+    >>> cf.fit_transform(X, y)      # doctest: +NORMALIZE_WHITESPACE
+    array([[ 0.,  1.,  3.],
+           [ 0.,  1.,  3.],
+           [ 0.,  1.,  3.],
+           [ 0.,  1.,  3.],
+           [ 1.,  4.,  0.],
+           [ 1.,  4.,  0.],
+           [ 1.,  4.,  0.],
+           [ 1.,  4.,  0.]])
+
```

Comment:
explicitly note that the last two columns correspond to y=0 and y=1, while the first column duplicates the input data.